include/mun/runtime.h

```diff
@@ -0,0 +1,275 @@
+         *
+         * \param handle a runtime handle
+         */
+        Runtime(MunRuntimeHandle handle) noexcept
+            : m_handle(handle)
+        {}
+
+    public:
+        /** Move constructs a runtime
+         *
+         * \param other an rvalue reference to a runtime
+         */
+        Runtime(Runtime&& other) noexcept
+            : m_handle(other.m_handle) {
+            other.m_handle._0 = nullptr;
+        }
+
+        /** Destructs a runtime */
+        ~Runtime() noexcept {
+            mun_runtime_destroy(m_handle);
+        }
+
+        /** Retrieves `FunctionInfo` from the runtime for the corresponding `fn_name`.
+         *
+         * \param fn_name the name of the desired function
+         * \param fn_info an output reference to an optional `MunFunctionInfo` struct
+         * \return a possible error
+         */
+        Error get_function_info(std::string_view fn_name, std::optional<MunFunctionInfo>& fn_info) noexcept {
+            bool has_fn;
+            MunFunctionInfo temp;
+            auto error = Error(mun_runtime_get_function_info(m_handle, fn_name.data(), &has_fn, &temp));
+            if (!error) {
+                if (has_fn) {
+                    fn_info.emplace(std::move(temp));
+                } else {
+                    fn_info.reset();
+                }
+            }
+
+            return error;
+        }
+
+        /** Checks for updates to hot reloadable assemblies.
+         *
+         * \param updated an output reference signalling whether the runtime was updated
+         * \return a possible error
+         */
+        Error update(bool& updated) {
+            return Error(mun_runtime_update(m_handle, &updated));
+        }
+
+    private:
+        MunRuntimeHandle m_handle;
+    };
+
+    /** Construct a new runtime that loads the library at `library_path` and its dependencies.
+     *
+     * On failure, the error is logged to the error log.
+     *
+     * \param library_path the path to a Mun library
+     * \return an optional runtime
+    .*/
+    std::optional<Runtime> make_runtime(std::string_view library_path) noexcept {
+        MunRuntimeHandle handle;
+        if (auto error = Error(mun_runtime_create(library_path.data(), &handle))) {
+            std::cerr << "Failed to construct Mun runtime due to error: " << error.message() << std::endl;
+            return std::nullopt;
+        }
+
+        return Runtime(handle);
+    }
+
+    /** Invokes the runtime function corresponding to `fn_name` with arguments `args`.
+     *
+     * \param runtime the runtime
+     * \param fn_name the name of the desired function
+     * \param args zero or more arguments to supply to the function invocation
+     * \return an invocation result
+     */
+    template <typename Output, typename... Args>
+    InvokeResult<Output, Args...> invoke_fn(Runtime& runtime, std::string_view fn_name, Args... args) noexcept {
```

Comment:
Why is this a free function instead of a member of `Runtime`?

Reply:
The choice is inspired by this philosophy: https://www.drdobbs.com/cpp/how-non-member-functions-improve-encapsu/184401197